### PR TITLE
update: add about defining Kotlin compiler execution strategy

### PR DESCRIPTION
**Pull request by PetrakovichVictoria**

### CODE REVIEW

File: docs/topics/gradle.md [hunk 1]
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+
+_Kotlin compiler execution strategy_ is a set of characteristics describing where Kotlin compiler is executed and some other nuances.
+
+There are three compiler execution strategies:
+
+| Strategy       | Where Kotlin compiler is executed | Support for incremental compilation | Other characteristics                                                |
+|----------------|-----------------------------------|-------------------------------------|----------------------------------------------------------------------|
+| Daemon         | Inside its own daemon process     | Yes                                 | *Enabled by default*. Can be shared between different Gradle daemons |
+| In process     | Inside Gradle daemon process      | No                                  | May share heap with the Gradle daemon                                |
+| Out of process | In separate forked process        | No                                  | -                                                                    |
```
Comment (ALikhachev) [inline docs/topics/gradle.md:982]:
```suggestion
| Out of process | In a separate forked process        | No                                  | -                                                                    |
```

File: docs/topics/gradle.md [hunk 2]
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+## Defining Kotlin compiler execution strategy
+
+_Kotlin compiler execution strategy_ is a set of characteristics describing where Kotlin compiler is executed and some other nuances.
+
+There are three compiler execution strategies:
+
+| Strategy       | Where Kotlin compiler is executed | Support for incremental compilation | Other characteristics                                                |
+|----------------|-----------------------------------|-------------------------------------|----------------------------------------------------------------------|
+| Daemon         | Inside its own daemon process     | Yes                                 | *Enabled by default*. Can be shared between different Gradle daemons |
+| In process     | Inside Gradle daemon process      | No                                  | May share heap with the Gradle daemon                                |
```
Comment (ALikhachev) [inline docs/topics/gradle.md:981]:
```suggestion
| In process     | Inside the Gradle daemon process      | No                                  | May share heap with the Gradle daemon                                |
```

File: docs/topics/gradle.md [hunk 3]
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
   >
   {type="note"}
 * If the `Xmx` is not specified, the Kotlin daemon will inherit it from the Gradle daemon.
+
+## Defining Kotlin compiler execution strategy
+
+_Kotlin compiler execution strategy_ is a set of characteristics describing where Kotlin compiler is executed and some other nuances.
```
Comment (ALikhachev) [inline docs/topics/gradle.md:974]:
```suggestion
_Kotlin compiler execution strategy_ defines a set of characteristics describing where Kotlin compiler is executed and some other nuances.
```

**Review by ALikhachev — CHANGES_REQUESTED** [review]

File: docs/topics/gradle.md [hunk 4]
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+
+The available values for `kotlin.compiler.execution.strategy` properties (both system and Gradle's) are:
+1. `in-process`
+2. `ouf-of-process`
+3. `daemon` (default)
+
+The available values for the `compilerExecutionStrategy` task property are:
+1. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.IN_PROCESS`
+2. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.OUT_OF_PROCESS`
+3. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.DAEMON` (default)
```
Comment (Tapchicoma) [inline docs/topics/gradle.md:999]: Probably swap it with 1. position

File: docs/topics/gradle.md [hunk 5]
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+* The task property `compilerExecutionStrategy` takes priority over the system and the Gradle properties `kotlin.compiler.execution.strategy`.
+* The Gradle property takes priority over the system property.
+
+The available values for `kotlin.compiler.execution.strategy` properties (both system and Gradle's) are:
+1. `in-process`
+2. `ouf-of-process`
+3. `daemon` (default)
+
+The available values for the `compilerExecutionStrategy` task property are:
+1. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.IN_PROCESS`
```
Comment (PetrakovichVictoria) [inline docs/topics/gradle.md:997]:
```suggestion
1. `org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy.DAEMON` (default)
```

**Comment by PetrakovichVictoria** [comment]
@koshachy @p7nov I need a tech writers' review.
I fixed all the remarks from the tech review.
The text is pretty much the same as in the "What's new", just added the definition of the Kotlin compiler strategy.

File: docs/topics/gradle.md [hunk 6]
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+kotlin.compiler.execution.strategy=out-of-process
+```
+
+Use the task property `compilerExecutionStrategy` in the `build.gradle.kts` buildscript:
+
+```kotlin
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy
+
+// Some code
```
Comment (p7nov) [inline docs/topics/gradle.md:1013]: No "some code" please. Either check how we show Gradle scripts with imports in other examples on this page or consider using FQNs in the code.

**Review by p7nov — CHANGES_REQUESTED** [review]
The part after the table is written in the What's new style, which isn't suitable for docs. Please rewrite it so that it isn't tied to the specific release. You can mention old things and versions to preserve compatibility, but do it in some isolated parts so that you can remove them instantly when these things are deprecated.
The doc should be "timeless" and relevant no matter when people read it.

File: docs/topics/gradle.md [hunk 7]
```diff
@@ -968,3 +968,51 @@ When configuring the Kotlin daemon's JVM arguments, note that:
   >
   {type="note"}
 * If the `Xmx` is not specified, the Kotlin daemon will inherit it from the Gradle daemon.
+
+## Defining Kotlin compiler execution strategy
+
+_Kotlin compiler execution strategy_ defines a set of characteristics describing where Kotlin compiler is executed and some other nuances.
```
[the remark this answers is not in the document]
Comment (PetrakovichVictoria) [reply docs/topics/gradle.md:974]: Agree about vague.

File: docs/topics/gradle.md [hunk 8]
```diff
@@ -968,3 +968,74 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+</tab>
+<tab title="Groovy" group-key="groovy">
+
+```groovy
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
+import org.jetbrains.kotlin.gradle.tasks.KotlinCompilerExecutionStrategy
+
+// ...
+
+tasks.withType(KotlinCompile.class)
```
Comment (PetrakovichVictoria) [inline docs/topics/gradle.md:1034]:
```suggestion
tasks.withType(KotlinCompile)
```
```suggestion
tasks.withType(KotlinCompile.class)
```

File: docs/topics/gradle.md [hunk 9]
```diff
@@ -968,3 +968,74 @@ When configuring the Kotlin daemon's JVM arguments, note that:
   >
   {type="note"}
 * If the `Xmx` is not specified, the Kotlin daemon will inherit it from the Gradle daemon.
+
+## Defining Kotlin compiler execution strategy
+
+_Kotlin compiler execution strategy_ describes where Kotlin compiler is executed and if incremental compilation is supported in each case.
```
Comment (p7nov) [inline docs/topics/gradle.md:974]: Maybe `defines`? Though the current wording seems OK too.

File: docs/topics/gradle.md [hunk 10]
```diff
@@ -968,3 +968,74 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+
+There are three compiler execution strategies:
+
+| Strategy       | Where Kotlin compiler is executed | Support for incremental compilation | Other characteristics                                                  |
+|----------------|-----------------------------------|-------------------------------------|------------------------------------------------------------------------|
+| Daemon         | Inside its own daemon process     | Yes                                 | *The default strategy*. Can be shared between different Gradle daemons |
+| In process     | Inside the Gradle daemon process  | No                                  | May share heap with the Gradle daemon                                  |
+| Out of process | In a separate forked process      | No                                  | -                                                                      |
```
Comment (p7nov) [inline docs/topics/gradle.md:982]: Not sure if `forked` adds meaning here. Does it have any practical use?

Comment (p7nov) [reply docs/topics/gradle.md:982]: For me, it only raised questions "Forked from what? What should I do with the fact that it was forked?"

File: docs/topics/gradle.md [hunk 11]
```diff
@@ -968,3 +968,74 @@ When configuring the Kotlin daemon's JVM arguments, note that:
...
+
+To define a Kotlin compiler execution strategy, you can use one of the following properties:
+* The `kotlin.compiler.execution.strategy` Gradle property.
+* The `compilerExecutionStrategy` compile task property.
+* The `-Dkotlin.compiler.execution.strategy` system property, which will be removed in future releases.
```
Comment (p7nov) [inline docs/topics/gradle.md:987]: Let's put the last thought in a separate note. Also, is it deprecated in 1.6.20? If yes, say this explicitly.

**Review by p7nov — APPROVED** [review]
Fine by me now, a couple of minor comments.

File: docs/topics/gradle.md (hunk 9 above)
Comment (PetrakovichVictoria) [inline docs/topics/gradle.md:974]:
```suggestion
_Kotlin compiler execution strategy_ defines where Kotlin compiler is executed and if incremental compilation is supported in each case.
```

Comment (PetrakovichVictoria) [reply docs/topics/gradle.md:974]: Defines is better indeed, thank you.

File: docs/topics/gradle.md (hunk 10 above)
Comment (PetrakovichVictoria) [reply docs/topics/gradle.md:982]: @ALikhachev could you help, please?

File: docs/topics/gradle.md (hunk 11 above)
Comment (PetrakovichVictoria) [reply docs/topics/gradle.md:987]: >is it deprecated in 1.6.20?

I don't have such info, @ALikhachev could you help, please?

Comment (ALikhachev) [reply docs/topics/gradle.md:987]: It's not deprecated yet, and we don't show any warnings on it's usages. We will add deprecation warning in 1.7 and then remove it in 1.8.

File: docs/topics/gradle.md (hunk 10 above)
Comment (ALikhachev) [reply docs/topics/gradle.md:982]: Yep, I guess we can remove `forked`. It's a technical detail
I'd suggest to rephrase it to be more clear
like
```suggestion
| Out of process | In a separate process for each call      | No                                  | -                                                                      |
```
So the new description emphasises that a new process will be created for each compilation task.
